### PR TITLE
[REF] Move membership date calc from v3 api to BAO

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
[REF] Move membership date calc from v3 api to BAO 

Before
----------------------------------------
code to calculated membership dates  for new membership is in the v3 api and in every form that creates memberships

After
----------------------------------------
calculated in the BAO for v3 api & membership batch entry form

Technical Details
----------------------------------------
This moves code to calculate membership dates for new memberships from the
v3 api Membership.create and the batch entry membership form to the membership bao.

I think more cleanup could follow this - in the api most obviously and
the BAO is a bit insane still

This has test cover in the v3 api membership test and in
CRM_Batch_Form_EntryTest

Comments
----------------------------------------

### CODE REVIEW

**Comment by mattwire** [comment]
```
CRM_Contribute_BAO_ContributionRecurTest::testAutoRenewalWhenOneMemberIsDeceased
End date incorrect.
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'2021-09-27'
+'2021-09-23'

/home/jenkins/bknix-dfl/build/core-20397-2n470/web/sites/all/modules/civicrm/tests/phpunit/CRM/Contribute/BAO/ContributionRecurTest.php:392
/home/jenkins/bknix-dfl/build/core-20397-2n470/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:236
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671
```

**Comment by eileenmcnaughton** [comment]
I took a look at that test - it's going to take a bit to make sense of it. If I don't find time to do that in the next few days I'll close this until I can find time. It's doing some kinda artificial stuff which might not really be valid but I have to dig

**Comment by eileenmcnaughton** [comment]
also - it will pass in 2 days time - for 24 hours - I think

**Comment by eileenmcnaughton** [comment]
Yeah - I'm not gonna look again this week - I'll close for now  - I have some reviewable PRs on cleanup on membership in the batch entry space still open

File: CRM/Member/BAO/Membership.php [hunk 1]
```diff
@@ -275,9 +275,23 @@ public static function create(&$params, $ids = []) {
...
-      $params['join_date'] = $params['join_date'] ?: 'null';
+      if (empty($params['id'])
+        && (empty($params['start_date']) || empty($params['join_date']) || (empty($params['end_date']) && !$isLifeTime))) {
+        // This is a new membership, calculate the membership dates.
+        $calcDates = CRM_Member_BAO_MembershipType::getDatesForMembershipType(
+          $params['membership_type_id'],
+          $params['join_date'] ?? NULL,
+          $params['start_date'] ?? NULL,
+          $params['end_date'] ?? NULL,
+          $params['num_terms'] ?? 1
```
Comment (eileenmcnaughton) [inline CRM/Member/BAO/Membership.php:286]: check this param name - membership number terms

**Pull request closed without merging**